### PR TITLE
Add new warning for deprecated keysyms

### PR DESCRIPTION
**Pull request by wismill**
Add new warning for deprecated keysyms. Useful in particular for _xkeyboard-config_.

TODO:
- [x] Perf?
- [ ] Public API?
- [x] Changelog entry

### CODE REVIEW

**Comment by bluetech** [comment]
I am doubtful on the cost vs. benefit of warning about deprecated keysym names. The deprecated names are never going to be removed. And more generally, basically all of the named keysyms are deprecated in favor of Unicode keysyms...

Maybe xkeyboard-config wants to avoid using deprecated names? If so, maybe do it as a lint script directly in xkeyboard-config?

**Comment by wismill** [comment]
> I am doubtful on the cost vs. benefit

Are we talking about perf, maintenance, both, other? About perf, I thought we could use a flag to make it opt-in. About maintenance: I think there is no maintenance issue, as long as the conventions to deprecate keysyms hold. The table is generated in the same script that update keysyms name lookup.

> The deprecated names are never going to be removed. 

Yes [they are](https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/68). Well the patch has been reverted since, but the plan is to delete them.

> And more generally, basically all of the named keysyms are deprecated in favor of Unicode keysyms...

This is not the policy in `xkeyboard-config`, where non-deprecated named keysyms are preferred over Unicode ones.

> Maybe xkeyboard-config wants to avoid using deprecated names? If so, maybe do it as a lint script directly in xkeyboard-config?

This is the case. But there is also the case for people developing their own layouts.

**Comment by wismill** [comment]
I think it would be enough to check for deprecated keysyms only with a special version dev version of `compile-keymap`; thus keeping the published version fast. The dev version would be only used in CI, e.g. for `xkeyboard-config`.

**Comment by wismill** [comment]
Rebased. Testing keysym deprecation is now only done when verbosity is ≥ 10.

**Comment by wismill** [comment]
Deprecation test is now done only when the new flag `XKB_CONTEXT_EXHAUSTIVE_CHECKS` is set. Without it, the new benchmark `bench-compile-keymap` gives only about +0,2% more time that on master. With the test, it take about +0,8%. I think this is reasonable!

**Review by bluetech — COMMENTED** [review]
>> The deprecated names are never going to be removed.
>
> Yes [they are](https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/68). Well the patch has been reverted since, but the plan is to delete them.

I'm surprised a breaking change was attempted. IMO it's better to leave the deprecated keysyms. But if xorgproto is going to remove them then I agree it would be good to warn about them.

> Deprecation test is now done only when the new flag XKB_CONTEXT_EXHAUSTIVE_CHECKS is set. Without it, the new benchmark bench-compile-keymap gives only about +0,2% more time that on master. With the test, it take about +0,8%. I think this is reasonable!

I thought the verbosity guard was pretty good myself, and allows the user/keymap auther to control it. Why is a context flag better?

File: src/keysym.c [hunk 1]
```diff
@@ -264,8 +264,7 @@ xkb_keysym_is_deprecated(xkb_keysym_t keysym,
                          const char *name,
                          const char **reference_name)
 {
-    /* FIXME: update this code once #357 is merged using XKB_KEYSYM_MAX */
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/keysym.c]: Fixed

File: src/keysym.h [hunk 2]
```diff
@@ -63,7 +63,7 @@ enum xkb_keysym_format {
     XKB_KEYSYM_FORMAT_NONE = 0,
     XKB_KEYSYM_FORMAT_NAME,
     XKB_KEYSYM_FORMAT_UNICODE,
-    XKB_KEYSYM_FORMAT_HEXADECIMAL
+    XKB_KEYSYM_FORMAT_NUMERIC
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/keysym.h:66]: Fixed

**Comment by wismill** [comment]
Rebased, cleaned commit history.

File: scripts/makekeys [hunk 3]
```diff
@@ -1,15 +1,77 @@
...
+
+
+@dataclass
+class Keysym:
+    name: str
+    value: int
+    deprecated: Deprecation
+
+
+def parse_keysyms(path: Path):
```
[the remark this answers is not in the document]
Comment (wismill) [reply scripts/makekeys]: done

File: scripts/makekeys [hunk 4]
```diff
@@ -1,15 +1,77 @@
...
+                    name=m.group("name"),
+                    value=int(m.group("value"), 16),
+                    deprecated=Deprecation.EXPLICIT
+                    if m.group("deprecated") or m.group("unicode") or
+                       m.group("name") in EXTRA_DEPRECATED_KEYSYMS
+                    else Deprecation.NONE,
+                )
+
+
+def get_keysyms(path: Path):
```
[the remark this answers is not in the document]
Comment (wismill) [reply scripts/makekeys]: done

File: scripts/makekeys [hunk 5]
```diff
@@ -1,15 +1,77 @@
...
+EXTRA_DEPRECATED_KEYSYMS = (
+    "Ext16bit_L",
+    "Ext16bit_R"
+)
+
+
+class Deprecation(Enum):
+    NONE = "none"
+    EXPLICIT = "explicit"
+    IMPLICIT = "implicit"
```
[the remark this answers is not in the document]
Comment (wismill) [reply scripts/makekeys]: done

**Comment by wismill** [comment]
Valgrind detected a bug that does not trigger on my computer. Will check it later.

**Comment by wismill** [comment]
Run locally in podman without error; re-run the failed jobs and no error 🤔

**Comment by wismill** [comment]
Rebased and improved benchmark.

**Pull request merged**